### PR TITLE
[Foundation] Do not get in an infinite loop with empty creds.

### PR DESCRIPTION
**Pull request by mandel-macaque**
As described in the issue, when the credentials for the base auth are set
to be empty the request reaches a timeout.

The inner issue is that when we have empty credentials the query that
is performed to the credentials db of the system throws an exception.
This exception is not raised to our code, but simply prints out in stderr.

When this situation occurs we get in an infinite loop in which we keep
sending the same credentials, we get another challenge, but the OS does
not do the right thing, we believe that we have to provide the creds
again (which are the same) and we get another challenge.. this goes on
an on until we reach the timeout.

Fortunately we know the number of failed challenges and the failure
response. With this information we can deduce if we already sent the
credentials, and if we did, do not set them again. In this case, the OS
sees no credentials in out delegate, uses the default handler and
correctly returns a 401.

fixes: https://github.com/xamarin/xamarin-macios/issues/8342
fixes: https://github.com/xamarin/xamarin-macios/issues/8344

The following attached app allows you to test the following:

1. Empty creds to the url provided by the user.
2. Test basic auth works (comment the first creds, uncomment the second ones, uncomment the request to httpbin.org)
3. Test basic auth works with the wrong creds (similar to 2 but uncomment the wrong creds).
[TestiOSHttpClient.zip](https://github.com/xamarin/xamarin-macios/files/4594666/TestiOSHttpClient.zip)

### CODE REVIEW

File: src/Foundation/NSUrlSessionHandler.cs [hunk 1]
```diff
@@ -757,7 +757,23 @@ public override void DidReceiveChallenge (NSUrlSession session, NSUrlSessionTask
...
+					// ptcl = htps;
+					// "r_Attributes" = 1;
+					// sdmn = test;
+					// srvr = "jigsaw.w3.org";
+					// sync = syna;
+					// }
+					// do remember that we ALWAYS get a challange, so the failure count has to be 1 or more for this check, 1 would be the first time
+					var nsurlRespose = challenge.FailureResponse as NSHttpUrlResponse;
+					var responseIsUnauthorized = (nsurlRespose == null) ? false : nsurlRespose.StatusCode == (int) HttpStatusCode.Unauthorized && challenge.PreviousFailureCount > 0;
+					if (authType != RejectProtectionSpaceAuthType && !responseIsUnauthorized) {
```
Comment (spouliot) [inline src/Foundation/NSUrlSessionHandler.cs:776]: nitpick: if `authType == RejectProtectionSpaceAuthType` then you're computing all this new stuff for nothing
might be better to nest the new code under the existing `if (authType != RejectProtectionSpaceAuthType) {`

File: src/Foundation/NSUrlSessionHandler.cs [hunk 2]
```diff
@@ -757,7 +757,23 @@ public override void DidReceiveChallenge (NSUrlSession session, NSUrlSessionTask
 
 				if (sessionHandler.Credentials != null && TryGetAuthenticationType (challenge.ProtectionSpace, out string authType)) {
 					NetworkCredential credentialsToUse = null;
-					if (authType != RejectProtectionSpaceAuthType) {
+					// interesting situation, when we use a credential that we created that is empty, we are not getting the RejectProtectionSpaceAuthType,
+					// nevertheless, we need to check is not the first challange we will continue trusting the
```
Comment (whitneyschmidt) [inline src/Foundation/NSUrlSessionHandler.cs:761]: nit: `challenge` `status response`

File: src/Foundation/NSUrlSessionHandler.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Foundation/NSUrlSessionHandler.cs:776]: good point

**Comment by mandel-macaque** [comment]
@monojenkins backport d16-7

**Pull request merged**